File: ios/Classes/OpenFilePlugin.m
#import "OpenFilePlugin.h"

@interface OpenFilePlugin ()<UIDocumentInteractionControllerDelegate>
@end

static NSString *const CHANNEL_NAME = @"open_file";

@implementation OpenFilePlugin{
    FlutterResult _result;
    UIViewController *_viewController;
    UIDocumentInteractionController *_documentController;
    UIDocumentInteractionController *_interactionController;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:CHANNEL_NAME
                                     binaryMessenger:[registrar messenger]];
    UIViewController *viewController =
    [UIApplication sharedApplication].delegate.window.rootViewController;
    OpenFilePlugin* instance = [[OpenFilePlugin alloc] initWithViewController:viewController];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)initWithViewController:(UIViewController *)viewController {
    self = [super init];
    if (self) {
        _viewController = viewController;
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"open_file" isEqualToString:call.method]) {
        _result = result;
        NSString *msg = call.arguments[@"file_path"];
        if(msg==nil){
            NSDictionary * dict = @{@"message":@"the file path cannot be null", @"type":@-4};
            NSData * jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:nil];
            NSString * json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
            result(json);
            return;
        }
        NSFileManager *fileManager=[NSFileManager defaultManager];
        BOOL fileExist=[fileManager fileExistsAtPath:msg];
        if(fileExist){
            //            NSURL *resourceToOpen = [NSURL fileURLWithPath:msg];
//            NSString *exestr = [[msg pathExtension] lowercaseString];
            _documentController = [UIDocumentInteractionController interactionControllerWithURL:[NSURL fileURLWithPath:msg]];
            _documentController.delegate = self;
            NSString *uti = call.arguments[@"uti"];
            BOOL isBlank = [self isBlankString:uti];
            if(!isBlank){
                _documentController.UTI = uti;
            }
//             else{
//                 if([exestr isEqualToString:@"rtf"]){
//                     _documentController.UTI=@"public.rtf";
//                 }else if([exestr isEqualToString:@"txt"]){
//                     _documentController.UTI=@"public.plain-text";
//                 }else if([exestr isEqualToString:@"html"]||
//                          [exestr isEqualToString:@"htm"]){
//                     _documentController.UTI=@"public.html";
//                 }else if([exestr isEqualToString:@"xml"]){
//                     _documentController.UTI=@"public.xml";
//                 }else if([exestr isEqualToString:@"tar"]){
//                     _documentController.UTI=@"public.tar-archive";
//                 }else if([exestr isEqualToString:@"gz"]||
//                          [exestr isEqualToString:@"gzip"]){
//                     _documentController.UTI=@"org.gnu.gnu-zip-archive";
//                 }else if([exestr isEqualToString:@"tgz"]){
//                     _documentController.UTI=@"org.gnu.gnu-zip-tar-archive";
//                 }else if([exestr isEqualToString:@"jpg"]||
//                          [exestr isEqualToString:@"jpeg"]){
//                     _documentController.UTI=@"public.jpeg";
//                 }else if([exestr isEqualToString:@"png"]){
//                     _documentController.UTI=@"public.png";
//                 }else if([exestr isEqualToString:@"avi"]){
//                     _documentController.UTI=@"public.avi";
//                 }else if([exestr isEqualToString:@"mpg"]||
//                          [exestr isEqualToString:@"mpeg"]){
//                     _documentController.UTI=@"public.mpeg";
//                 }else if([exestr isEqualToString:@"mp4"]){
//                     _documentController.UTI=@"public.mpeg-4";
//                 }else if([exestr isEqualToString:@"3gpp"]||
//                          [exestr isEqualToString:@"3gp"]){
//                     _documentController.UTI=@"public.3gpp";
//                 }else if([exestr isEqualToString:@"mp3"]){
//                     _documentController.UTI=@"public.mp3";
//                 }else if([exestr isEqualToString:@"zip"]){
//                     _documentController.UTI=@"com.pkware.zip-archive";
//                 }else if([exestr isEqualToString:@"gif"]){
//                     _documentController.UTI=@"com.compuserve.gif";
//                 }else if([exestr isEqualToString:@"bmp"]){
//                     _documentController.UTI=@"com.microsoft.bmp";
//                 }else if([exestr isEqualToString:@"ico"]){
//                     _documentController.UTI=@"com.microsoft.ico";
//                 }else if([exestr isEqualToString:@"doc"]){
//                     _documentController.UTI=@"com.microsoft.word.doc";
//                 }else if([exestr isEqualToString:@"xls"]){
//                     _documentController.UTI=@"com.microsoft.excel.xls";
//                 }else if([exestr isEqualToString:@"ppt"]){
//                     _documentController.UTI=@"com.microsoft.powerpoint.​ppt";
//                 }else if([exestr isEqualToString:@"wav"]){
//                     _documentController.UTI=@"com.microsoft.waveform-​audio";
//                 }else if([exestr isEqualToString:@"wm"]){
//                     _documentController.UTI=@"com.microsoft.windows-​media-wm";
//                 }else if([exestr isEqualToString:@"wmv"]){
//                     _documentController.UTI=@"com.microsoft.windows-​media-wmv";
//                 }else if([exestr isEqualToString:@"pdf"]){
//                     _documentController.UTI=@"com.adobe.pdf";
//                 }else {
//                     NSLog(@"doc type not supported for preview");
//                     NSLog(@"%@", exestr);
//                 }
//             }
            @try {
                BOOL previewSucceeded = [_documentController presentPreviewAnimated:YES];
                if(!previewSucceeded){
                    [_documentController presentOpenInMenuFromRect:CGRectMake(500,20,100,100) inView:[UIApplication sharedApplication].delegate.window.rootViewController.view animated:YES];
                }
            }@catch (NSException *exception) {
                NSDictionary * dict = @{@"message":@"File opened incorrectly。", @"type":@-4};
                NSData * jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:nil];
                NSString * json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
                result(json);
            }
        }else{
            NSDictionary * dict = @{@"message":@"the file does not exist", @"type":@-2};
            NSData * jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:nil];
            NSString * json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

            result(json);
        }
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)documentInteractionControllerDidEndPreview:(UIDocumentInteractionController *)controller {
    NSDictionary * dict = @{@"message":@"done", @"type":@0};
    NSData * jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:nil];
    NSString * json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

    _result(json);
}

- (void)documentInteractionControllerDidDismissOpenInMenu:(UIDocumentInteractionController *)controller {
      NSDictionary * dict = @{@"message":@"done", @"type":@0};
      NSData * jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:nil];
      NSString * json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

      _result(json);
}

- (UIViewController *)documentInteractionControllerViewControllerForPreview:(UIDocumentInteractionController *)controller {
    return [UIApplication sharedApplication].delegate.window.rootViewController;
}

- (BOOL) isBlankString:(NSString *)string {
    if (string == nil || string == NULL) {
        return YES;
    }
    if ([string isKindOfClass:[NSNull class]]) {
        return YES;
    }
    if ([[string stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] length]==0){
        return YES;
    }
    return NO;
}
@end

File: open_file_mac/macos/Classes/OpenFilePlugin.swift
import Cocoa
import FlutterMacOS

public class OpenFilePlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "open_file", binaryMessenger: registrar.messenger)
        let instance = OpenFilePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "open_file":
            let arguments = call.arguments as? Dictionary<String,Any>
            let filePath = arguments!["file_path"] as? String
            if(filePath==nil){
                self.result(result:result,message:"the file path cannot be null",type:-4)
                return
            }
            let fileExist = FileManager.default.fileExists(atPath: filePath!)
            if(fileExist){
                let documentURL = URL(fileURLWithPath: filePath!)
                let fileType = documentURL.pathExtension
                if(!canOpen(filePath: filePath!)){
                    self.openSelectPanel(filePath:filePath!,fileType: fileType ,result: result)
                }else{
                    self.open(documentURL: documentURL, result: result)
                }
            }else{
                self.result(result:result,message:"the file does not exist",type:-2)
            }
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func canOpen(filePath: String) ->Bool{
        return FileManager.default.isReadableFile(atPath: filePath)
    }
    
    private func convertDictionaryToString(dict:[String:Any]) -> String {
        var result:String = ""
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dict, options: JSONSerialization.WritingOptions.init(rawValue: 0))
            
            if let JSONString = String(data: jsonData, encoding: String.Encoding.utf8) {
                result = JSONString
            }
            
        } catch {
            result = ""
        }
        return result
    }
    
    private func openSelectPanel(filePath:String,fileType: String,result: @escaping FlutterResult){
        let fileUrl = URL(fileURLWithPath:filePath)
        let openPanel = NSOpenPanel()
        openPanel.canChooseFiles = false
        openPanel.canChooseDirectories = true
        openPanel.allowsMultipleSelection = false
        openPanel.allowedFileTypes = [fileType]
        openPanel.directoryURL = fileUrl
        openPanel.showsHiddenFiles = true
        openPanel.allowsOtherFileTypes = false
        let currentLanguage = Locale.current.languageCode
        openPanel.prompt = (currentLanguage == "zh") ?"允许":"Accept"
        openPanel.beginSheetModal(for: NSApplication.shared.mainWindow!) { (openResult) in
            if (openResult.rawValue == NSApplication.ModalResponse.OK.rawValue) {
                let selectedURL = openPanel.url!
                var isReadable: Bool = false
                if FileManager.default.fileExists(atPath: selectedURL.path, isDirectory: nil) && FileManager.default.isReadableFile(atPath: selectedURL.path) {
                    isReadable = true
                }
                if isReadable {
                    self.open(documentURL: fileUrl, result: result)
                } else {
                    self.result(result:result,message:"Operation not permitted",type:-3)
                }
            }else{
                self.result(result:result,message:"Operation not permitted",type:-3)
            }
        }
    }
    
    private func open(documentURL:URL,result: FlutterResult){
        NSWorkspace.shared.open([documentURL], withAppBundleIdentifier: nil, options: .default, additionalEventParamDescriptor: nil, launchIdentifiers: nil)
        self.result(result:result,message:"done",type:0)
    }
    
    private func result(result: FlutterResult,message:String,type:Int){
        let map = ["message":message, "type":type]as [String : Any]
        result(convertDictionaryToString(dict: map))
    }
    
    @available(*, deprecated, message: "This method is just test.")
    private func requestDiskPermission(onGranted:()->Void,onDenied:()->Void){
        
        if let url = URL(string: "x-apple.systempreferences:com.apple.preference.security?Privacy_AllFiles") {
            if NSWorkspace.shared.open(url) {
                print("Opened successfully")
            }else{
                onDenied()
                return
            }
        }
        
        let options: NSDictionary = [kAXTrustedCheckOptionPrompt.takeRetainedValue() as NSString: true]
        let status = AXIsProcessTrustedWithOptions(options)
        
        if (status != true) {
            print("Prompting for accessibility permissions")
            let alert = NSAlert()
            alert.messageText = "This app requires disk permissions to function properly. Please grant access in the System Preferences app when prompted."
            alert.addButton(withTitle: "OK")
            alert.runModal()
            onDenied()
            return
        }
        
        let task = Process()
        task.launchPath = "/usr/bin/sqlite3"
        task.arguments = ["/Library/Application Support/com.apple.TCC/TCC.db", "INSERT INTO access VALUES('kTCCServiceSystemPolicyAllFiles', '\(Bundle.main.bundleIdentifier!)', 0, 1, 0, NULL);"]
        task.launch()
        task.waitUntilExit()
        onGranted()
        print("Full Disk Access granted successfully")
        
    }
    
    func checkFullDiskAccess() -> Bool {
        let fileManager = FileManager.default
        let testFilePath = "/Library/Application Support/com.apple.TCC/TCC.db"
        
        if fileManager.isReadableFile(atPath: testFilePath) {
            return true
        } else {
            return false
        }
    }
    
    @available(*, deprecated, message: "This method is no longer used.")
    private func getUTI(fileType:String)->String{
        var uti = ""
        switch fileType.lowercased(){
        case "rtf":
            uti = "public.rtf"
        case "txt":
            uti = "public.plain-text"
        case "html","htm":
            uti = "public.html"
        case "xml":
            uti = "public.xml"
        case "tar":
            uti = "public.tar-archive"
        case "gz","gzip":
            uti = "org.gnu.gnu-zip-archive"
        case "tgz":
            uti = "org.gnu.gnu-zip-tar-archive"
        case "jpg","jpeg":
            uti = "public.jpeg"
        case "png":
            uti = "public.png"
        case "avi":
            uti = "public.avi"
        case "mpg","mpeg":
            uti = "public.mpeg"
        case "mp4":
            uti = "public.mpeg-4"
        case "3gpp","3gp":
            uti = "public.3gpp"
        case "mp3":
            uti = "public.mp3"
        case "zip":
            uti = "com.pkware.zip-archive"
        case "gif":
            uti = "com.compuserve.gif"
        case "bmp":
            uti = "com.microsoft.bmp"
        case "ico":
            uti = "com.microsoft.ico"
        case "doc":
            uti = "com.microsoft.word.doc"
        case "xls":
            uti = "com.microsoft.excel.xls"
        case "ppt":
            uti = "com.microsoft.powerpoint.​ppt"
        case "wav":
            uti = "com.microsoft.waveform-​audio"
        case "wm":
            uti = "om.microsoft.windows-​media-wm"
        case "wmv":
            uti = "om.microsoft.windows-​media-wmv"
        case "pdf":
            uti = "com.adobe.pdf"
        default:
            uti = ""
        }
        return uti
    }
}
